syntax = "proto3";

package hakoniwa;
option csharp_namespace = "HakoniwaGrpc";

service CoreService {
  //箱庭アセットを登録する
  rpc Register (AssetInfo) returns (NormalReply) {}
  //箱庭アセットを登録解除する
  rpc Unregister (AssetInfo) returns (NormalReply) {}

  //箱庭アセット非同期通知
  rpc AssetNotificationStart (AssetInfo) returns (stream AssetNotification) {}
  rpc AssetNotificationFeedback (AssetNotificationReply) returns (NormalReply) {}
}
enum AssetNotificationEvent {
  ASSET_NOTIFICATION_EVENT_START = 0;
  ASSET_NOTIFICATION_EVENT_END = 1;
}
enum ErrorCode {
  ERROR_CODE_OK = 0;
  ERROR_CODE_PERM = 1;
  ERROR_CODE_NOENT = 2;
}

message AssetNotification {
  AssetNotificationEvent event = 1;
}
message AssetNotificationReply {
  AssetNotificationEvent event = 1;
  AssetInfo asset = 2;
  ErrorCode ercd = 3;
}
message NormalReply {
  ErrorCode ercd = 1;
}

message AssetInfo {
  string name = 1;
}

message HAssetInfoList {
  repeated AssetInfo assets = 1;
}
